cmd/service-controller/controller.go

```diff
@@ -309,15 +320,24 @@ func (c *Controller) checkServiceFor(desired *ServiceBindings, actual *corev1.Se
 	if len(actual.Spec.Ports) > 0 {
 		if actual.Spec.Ports[0].Port != int32(desired.publicPort) {
 			update = true
+			if !hasOriginalPort(*actual) {
+				actual.ObjectMeta.Annotations[types.OriginalPortQualifier] = strconv.FormatInt(int64(actual.Spec.Ports[0].Port), 10)
+			}
 			actual.Spec.Ports[0].Port = int32(desired.publicPort)
 		}
 		if actual.Spec.Ports[0].TargetPort.IntValue() != desired.ingressPort {
 			update = true
+			if !hasOriginalTargetPort(*actual) {
+				actual.ObjectMeta.Annotations[types.OriginalTargetPortQualifier] = strconv.FormatInt(int64(actual.Spec.Ports[0].TargetPort.IntVal), 10)
+			}
 			actual.Spec.Ports[0].TargetPort = intstr.FromInt(desired.ingressPort)
 		}
 	}
 	if desired.headless == nil && !equivalentSelectors(actual.Spec.Selector, kube.GetLabelsForRouter()) {
 		update = true
+		if !hasOriginalSelector(*actual) {
+			actual.ObjectMeta.Annotations[types.OriginalSelectorQualifier] = utils.StringifySelector(actual.Spec.Selector)
+		}
```

Comment:
As above for targetPort, perhaps we should always just record the original selector when we change it.

Reply:
That is also correct, I will change it.